0000-interledger-tokens/0000-interledger-tokens.md

```diff
@@ -0,0 +1,242 @@
+
+It is good practice to change this secret regularly and to always issue tokens with a limited lifetime (a value of `exp` that is not too far in the future). In that way the issuer can be confident about when it can retire an old secret after it has been replaced, and reject any tokens that claim to be signed using that old secret.
+
+Whenever a new token is generated a new signing and encryption key should be derived for this token.
+
+The ILSP should generate a unique identifier for the token, the `jti` (a 128-bit UUID is recommended) and derive the keys as follows:
+
+```
+token_seed = SHA-256(payer_secret || jti)
+signing_key = SHA-256("signing_key" || token_seed)
+encryption_key = SHA256("encryption_key" || token_seed)
+```
+
+Any restrictions required by the payer are defined in Payer Data which is encrypted into a JWE using the derived `encryption_key` as the CEK per the instructions in RFC7516. 
+
+The value of the `alg` header must be `dir` and the value of the `enc` header must be `A256GCM` in the JWE Protected Header of this JWE. No other headers are allowed.
+
+The 96-bit IV (the value of the ``) must be randomly generated.
+
+The JWE, using the Flattened JWE JSON Serialization Syntax as described in [RFC7516 Section 7.2.2]((https://tools.ietf.org/html/rfc7516#section-7.2.2), is used as the value of the `payer` claim in the token.
+
+Any restrictions specified by the payee are defined in the Payee Data which is set as the value of the `payee` claim.
+
+The unique JWT identifier (used in the key generation process) is used for the value of the `jti` claim, the ILP Address of the payee as the value of the `sub` claim and an identifer of the payee is specified in the `iss` claim. If the token is generated by an ILSP (other than the payer) then the value of `iss` must uniquely identify the payer.
+
+The `aud` claim must be an HTTPS URL where the payee is able to submit the token and request a payment.
+
+If the token is to be time limited then either an `exp` or a `nbf` or both claims must be set. It is recommended that an `exp` value is always set and that issuers enforce a policy of keeping this window short to allow for old secrets to be retired.
+
+The `iat` value should be set to the current time as this will assist in identifying the `sender_secret` that was used (in combination with the `iss` value) assuming the issuer regularly changes the `sender_secret`.
+
+The token should then be signed using the `HMAC using SHA-256` algorithm as described in [RFC7515](https://tools.ietf.org/html/rfc7515) and serialized using the Flattened JWS JSON Serialization Syntax as defined in [RFC7515 Section 7.2.2](https://tools.ietf.org/html/rfc7515#section-7.2.2).
+
+No unprotected headers are allowed.
+
+### Redemption
+
+A token is redeemed by submitting it to the endpoint defined in the `aud` value. An ILSP that is processing a token should reject it if the value of `aud` does not match the URL of the endpoint at which the token was submitted.
```

Comment:
I'm not a fan of using a completely different protocol just for this use case. I think we could enable this kind of flexible pull-payment with either a payment pointer and version of SPSP that supports [duplex payment streams](https://github.com/interledger/rfcs/issues/388) or something like a BTP endpoint and token that lets you connect and send payments to yourself through that using any higher-level protocol you like.

Reply:
> something like a BTP endpoint and token that lets you connect and send payments to yourself through that using any higher-level protocol you like.

That's exactly what this is except that it's a more sophisticated token because it's a bearer token with the limits encoded in the token itself that can be generated by an entirely offline sender. That is a critical differentiator.

Reply:
Okay, but then it shouldn't have a different way to redeem it. If you just use BTP with a bearer token, the fact that you're using a fancy token that encodes the rules will be completely opaque to the merchant and your BTP server can support whatever token formats it wants.

Reply:
The purpose of a standard is to not tie the token format to the protocol. It's doesn't have to be BTP and in many cases it will be desirable not to be. Such as when the entity redeeming the token has no idea what BTP is they simply have a token they want to submit to a REST API somewhere and get a response to say they got paid.

> BTP server can support whatever token formats it wants.

Sure, if my BTP server is the only thing that can issue the tokens for me. i.e. That doesn't support an offline sender.

> the fact that you're using a fancy token that encodes the rules will be completely opaque to the merchant

Not at all. The token is intentionally not encrypted but just signed. The merchant can look at the token and determine immediately if they are likely to be able to redeem it and for how much.

Reply:
Unless you're using a payment pointer, which includes protocol negotiation in it, I don't see how you're not tied to a protocol for actually redeeming the money.

> they simply have a token they want to submit to a REST API somewhere and get a response to say they got paid.

The receiver also needs to be running a receiver that speaks the same application layer protocol as the payer's ILSP.

> Sure, if my BTP server is the only thing that can issue the tokens for me. i.e. That doesn't support an offline sender.

That's not true, and I implemented this in "ILP3" with macaroons. If the BTP server uses a token format that allows for offline generation or caveating, the sender can be offline and generate sub-tokens that the BTP server will honor.

The benefit of using bearer tokens that are opaque to the merchant is that ILSPs can support different formats and ever more complex rules without the merchant changing their flow at all.

Reply:
I think payment pointers are preferable to BTP because they make use of a layer on top of ILP instead of a layer underneath it. But yeah, a payment pointer is also capable of containing identifying information so that the ILSP can tell what merchant is pulling.

Reply:
> The receiver also needs to be running a receiver that speaks the same application layer protocol as the payer's ILSP.

The current proposal allows the payee to submit the application data so they are defining the application layer protocol on both ends.

> That's not true, and I implemented this in "ILP3" with macaroons. If the BTP server uses a token format that allows for offline generation or caveating, the sender can be offline and generate sub-tokens that the BTP server will honor.

Then that was just another format for exactly the same thing. Macaroons instead of JWTs.

> The benefit of using bearer tokens that are opaque to the merchant is that ILSPs can support different formats and ever more complex rules without the merchant changing their flow at all.

The merchant doesn't HAVE to change their flow. If the merchant has the ability to read some data in the token then they have the advantage of being able to change their flow if they choose to.

The benefit of using tokens that have some opaque and some visible data is that the bearer knows what they are actually carrying. If I get a token for an ILP Address I no longer own or that has expired I know that before trying to redeem it.

Reply:
> I think payment pointers are preferable to BTP because they make use of a layer on top of ILP instead of a layer underneath it.

I don't really follow the on top vs underneath thing.

We should be very wary of payment pointers as a way to pull payments. This is exactly the mistake that was made decades ago with cards. All you needed was someone's card number and you could start pulling money out of their account. The result is layers and layers of complexity to try and shield card numbers from being shared.

The purpose of payment pointers is to give someone a pointer to an account they can pay into not the other way around.

Reply:
> I don't really follow the on top vs underneath thing.

using BTP for application layer concerns means everyone needs to use it as the communication under ILP. Using a protocol like SPSP that rides on top of ILP means that everyone only needs to have ILP in common. It's like how lightning uses the lnd RPC endpoint in their application level invoice flow, which makes it harder to become interoperable.

> This is exactly the mistake that was made decades ago with cards.

Not sure what the difference is between having a token get leaked vs. having a payment pointer get leaked. Either way, they're authorized for a specific amount of money/throughput and can be revoked.